### ISSUE TITLE
Error when trying to run two-particles example

### ISSUE THREAD

**Issue by ovgeorge** [anders-dc/Granular.jl#11]
> ➜  gran julia -v
> julia version 1.0.1
> ➜  gran julia two-grains.jl
> [ Info: Added grain 1
> [ Info: Added grain 2
> [ Info: Time step length t=8.478741928617433e-5 s
> 
> ERROR: LoadError: DimensionMismatch("tried to assign 2-element array to 3×1 destination")
> Stacktrace:
>  [1] throw_setindex_mismatch(::Array{Float64,1}, ::Tuple{Int64,Int64}) at ./indices.jl:169
>  [2] setindex_shape_check at ./indices.jl:221 [inlined]
>  [3] macro expansion at ./multidimensional.jl:641 [inlined]
>  [4] _unsafe_setindex!(::IndexLinear, ::Array{Float64,2}, ::Array{Float64,1}, ::UnitRange{Int64}, ::Int64) at ./multidimensional.jl:636
>  [5] _setindex! at ./multidimensional.jl:631 [inlined]
>  [6] setindex! at ./abstractarray.jl:998 [inlined]
>  [7] convertGrainDataToArrays(::Granular.Simulation) at /Users/george/.julia/packages/Granular/6XA3Z/src/grain.jl:527
>  [8] #writeGrainVTK#39(::Bool, ::Function, ::Granular.Simulation, ::String) at /Users/george/.julia/packages/Granular/6XA3Z/src/io.jl:348
>  [9] #writeGrainVTK at ./none:0 [inlined]
>  [10] #writeVTK#38(::String, ::Bool, ::Bool, ::Bool, ::Function, ::Granular.Simulation) at /Users/george/.julia/packages/Granular/6XA3Z/src/io.jl:317
>  [11] #writeVTK at ./logging.jl:0 [inlined]
>  [12] macro expansion at ./logging.jl:317 [inlined]
>  [13] #run!#3(::Bool, ::Int64, ::Bool, ::Bool, ::String, ::Bool, ::Function, ::Granular.Simulation) at /Users/george/.julia/packages/Granular/6XA3Z/src/simulation.jl:114
>  [14] run!(::Granular.Simulation) at /Users/george/.julia/packages/Granular/6XA3Z/src/simulation.jl:99
>  [15] top-level scope at none:0
>  [16] include at ./boot.jl:317 [inlined]
>  [17] include_relative(::Module, ::String) at ./loading.jl:1041
>  [18] include(::Module, ::String) at ./sysimg.jl:29
>  [19] exec_options(::Base.JLOptions) at ./client.jl:229
>  [20] _start() at ./client.jl:421
> in expression starting at /Users/george/polygon/gran/two-grains.jl:32
> ➜  gran julia two-grains.jl
> [ Info: Added grain 1
> [ Info: Added grain 2
> [ Info: Time step length t=8.478741928617433e-5 s
> 
> ERROR: LoadError: DimensionMismatch("tried to assign 2-element array to 3×1 destination")
> Stacktrace:
>  [1] throw_setindex_mismatch(::Array{Float64,1}, ::Tuple{Int64,Int64}) at ./indices.jl:169
>  [2] setindex_shape_check at ./indices.jl:221 [inlined]
>  [3] macro expansion at ./multidimensional.jl:641 [inlined]
>  [4] _unsafe_setindex!(::IndexLinear, ::Array{Float64,2}, ::Array{Float64,1}, ::UnitRange{Int64}, ::Int64) at ./multidimensional.jl:636
>  [5] _setindex! at ./multidimensional.jl:631 [inlined]
>  [6] setindex! at ./abstractarray.jl:998 [inlined]
>  [7] convertGrainDataToArrays(::Granular.Simulation) at /Users/george/.julia/packages/Granular/6XA3Z/src/grain.jl:527
>  [8] #writeGrainVTK#39(::Bool, ::Function, ::Granular.Simulation, ::String) at /Users/george/.julia/packages/Granular/6XA3Z/src/io.jl:348
>  [9] #writeGrainVTK at ./none:0 [inlined]
>  [10] #writeVTK#38(::String, ::Bool, ::Bool, ::Bool, ::Function, ::Granular.Simulation) at /Users/george/.julia/packages/Granular/6XA3Z/src/io.jl:317
>  [11] #writeVTK at ./logging.jl:0 [inlined]
>  [12] macro expansion at ./logging.jl:317 [inlined]
>  [13] #run!#3(::Bool, ::Int64, ::Bool, ::Bool, ::String, ::Bool, ::Function, ::Granular.Simulation) at /Users/george/.julia/packages/Granular/6XA3Z/src/simulation.jl:114
>  [14] run!(::Granular.Simulation) at /Users/george/.julia/packages/Granular/6XA3Z/src/simulation.jl:99
>  [15] top-level scope at none:0
>  [16] include at ./boot.jl:317 [inlined]
>  [17] include_relative(::Module, ::String) at ./loading.jl:1041
>  [18] include(::Module, ::String) at ./sysimg.jl:29
>  [19] exec_options(::Base.JLOptions) at ./client.jl:229
>  [20] _start() at ./client.jl:421
> in expression starting at /Users/george/polygon/gran/two-grains.jl:32
> 
>

**Comment by ovgeorge** [comment]
It is actually easy to fix, by adding zero third coordinate (unsure why, do Granular.js is actually doing three dimensional modelling? If so, why we have only cylindrical particles?), for example:

before
sim.grains[1].lin_vel = [1.0, 0.0]

after:
sim.grains[1].lin_vel = [1.0, 0.0, 0.0]

and change the whole code in this manner. This is the easy fix. Sadly it is nor working for 
double_gyre example. Probably the ocean needs to be changed, too, unsure how. 

I love the simplicity of the code and features it provides, too bad the project gives off a little abandoned vibe.

**Comment by anders-dc** [comment]
Hi, thanks for the heads up!
The examples are not part of the tests, so I rarely remember to check if they are up to date. I will run through them the next few days and get them up and running. The third dimension is there for pseudo-3d effects, and is normally not used. I have big plans for this code that will be put into reality the next year, so make sure to check it out later.

**Comment by ovgeorge** [comment]
Thanks for your reply! It is great to hear the project is not abandoned, because I have a big plans for it right now. Thank you for all the hard work! Are there any papers which I should cite when using your code?

**Comment by anders-dc** [comment]
Excellent! Please cite https://doi.org/10.1029/2018MS001299. Let me know if you run into any further problems as I am happy to help.

**Comment by anders-dc** [comment]
Hello again,
I am happy to report that all examples in `examples/` are now working in Julia 1.1. This is updated on the master branch on the development version, and in [release v0.4.3](
https://github.com/anders-dc/Granular.jl/releases/tag/v0.4.3) that will be available through the Julia package manager shortly.
Thanks again, Anders

**Comment by ovgeorge** [comment]
Andres, thank you again for great work! 

I have a small question, which I will ask here, so not to create too many threads. 

I'm trying to simulate granular gas with piston (I'm modelling piston as a East wall with constant speed boundary conditions). But, for some reason gas particles are moving very slow, compared to the wall even when I set their velocities to be high. Could you please take a look as say what is wrong with my code?


`#!/usr/bin/env julia
import Granular

sim = Granular.createSimulation(id="gas")
sim.ocean = Granular.createRegularOceanGrid([10, 10, 1], [1.0, 1.0, 1.0])

for i = 1:250
        Granular.addGrainCylindrical!(sim, [rand(), rand(), 0], 1e-2, 1e-2, lin_vel = [1e2*rand(), 1e2*rand(), 0])
end

#for i = 1:20
#    sim.grains[i].lin_vel = [1e1*rand(), 1e1*rand(), 0]
#end
Granular.fitGridToGrains!(sim, sim.ocean)

#Granular.setGridBoundaryConditions!(sim.ocean, "impermeable")
Granular.addWallLinearFrictionless!(sim, [0., 1., 0.], 0.0)
Granular.addWallLinearFrictionless!(sim, [0., 1., 0.], 1.0)
Granular.addWallLinearFrictionless!(sim, [1., 0., 0.], 0.0)
Granular.addWallLinearFrictionless!(sim, [1., 0., 0.], 1.0, bc="velocity", vel=-0.5)

# Set a velocity of 0.5 m/s along +x for the first grain, to make it bump into
# the second grain.

# Before we can run the simulation, we need to specify the computational time
# step, how often to produce output files for visualization, and for how long to
# run the simulation in model time [s]:
Granular.setTimeStep!(sim)
Granular.setOutputFileInterval!(sim, 1e-3)
Granular.setTotalTime!(sim, 0.1)

# Let's save the total kinetic energy before the simulation:
E_kin_before = Granular.totalGrainKineticTranslationalEnergy(sim)

# We can now run the simulation in a single call:
Granular.run!(sim)

# The kinetic energy after:
E_kin_after = Granular.totalGrainKineticTranslationalEnergy(sim)

# Report these values to console
@info "Kinetic energy before: $E_kin_before J"
@info "Kinetic energy after:  $E_kin_after J"

Granular.render(sim, animation=true)
`

**Comment by anders-dc** [comment]
Hello again, I just received notification about a comment to this issue, but maybe you deleted it?

Try adding the following lines after you add the grains:

```
for grain in sim.grains
    Granular.disableOceanDrag!(grain)
end
```

The ocean adds a viscous drag to any moving grains. The code was initially intended for plates og granular ice floating on the surface. However, it does not seem be a good idea to add viscous drag by default, so I might change that behavior.

**Comment by ovgeorge** [comment]
Thank you! I've accidentality deleted comment, thank you I will try it.

**Comment by anders-dc** [comment]
Hello again, just another thought. Currently you create grains at random positions within the start geometry. That can lead to stability problems if two grains are placed on top of each other with large overlap. I suggest doing something like the following to ensure no particles overlap at the start:

```
Granular.fitGridToGrains!(sim, sim.ocean)
for i = 1:250
    Granular.sortGrainsInGrid!(sim, sim.ocean)
    x_candidate = [rand(), rand(), 0.0]
    r_candidate = 1e-2
    if checkForContacts(sim, sim.ocean, x_candidate, r_candidate, return_when_overlap_found=true)
        Granular.addGrainCylindrical!(sim, x_candidate, r_candidate, 1e-2, lin_vel = [1e2*rand(),
1e2*rand(), 0])
    end
end
```

Feel free to reach out at anders@adamsgaard.dk

**Comment by ovgeorge** [comment]
@anders-dc Thank you, everything works! I have more questions, though. 

1. How to visualize walls added with addWallLinearFrictionless? 
2. How to make collisions with the wall completely elastic? Should I set wall's masses 1e10 or something like that?
3. I need grains to lose energy upon the impact with each others (but not the walls!), so I have to set
either
`contact_static_friction::Float64 = 0.4: contact static Coulomb frictional coefficient `
or
`contact_dynamic_friction::Float64 = 0.4: contact dynamic Coulomb frictional coefficient`

I think I need set the dynamic friction, correct? 

Which friction and elasticity models are used here?

**Comment by anders-dc** [comment]
>How to visualize walls added with addWallLinearFrictionless?

There are no built-in tools for this, but you can probably program it yourself in Paraview. I usually draw a box around the simulation domain, for example by importing the `ocean` grid VTK files.

>How to make collisions with the wall completely elastic? Should I set wall's masses 1e10 or something like that?

If the walls are set with `addWallLinearFrictionless!(..., bc, ..., contact_viscosity_normal, ...)` and `bc` is `"fixed"` or `"velocity"`, the wall mass does not matter as the wall does not move (`fixed`), or moves with a constant velocity (`velocity`). The default behavior is that interactions with free-moving grains are completely elastic as `contact_viscosity_normal=0` is the standard value in the `addWallLinearFrictionless` call.

>I need grains to lose energy upon the impact with each others (but not the walls!), so I have to set
either contact_static_friction::Float64 = 0.4: contact static Coulomb frictional coefficient 
or contact_dynamic_friction::Float64 = 0.4: contact dynamic Coulomb frictional coefficient
>I think I need set the dynamic friction, correct?

Yes, if the contact-frictional coefficients are small, grains will loose energy through contact-tangential sliding. A smaller contact-frictional value means that more energy is lost. I recommend setting both friction coefficients to the same value. If static friction is larger than dynamic friction, you will experience stick-slip at the contacts. If you want the grains to loose energy also from grain-normal collisions, you need to set `contact_viscosity_normal` for the grains.

>Which friction and elasticity models are used here?

For contact-normal and tangential components I use a Kelvin-Voigt model with linear elasticity and linear viscosity. If the viscosity is set to 0, which is the default, the viscous resistance is zero and the contacts become purely elastic. In all cases, the tangential force is limited through Coulomb slip. See eqs. 7 to 10 in doi 10.1029/2018MS001299 for the mathematical definitions.
